device_test.go

```diff
@@ -227,20 +217,54 @@ func (*failDetachLib) detachKernelDriver(h *libusbDevHandle, i uint8) error {
 
 func TestAutoDetachFailure(t *testing.T) {
 	t.Parallel()
-	fake := newFakeLibusb()
-	c := newContextWithImpl(&failDetachLib{fake})
-	defer c.Close()
-	dev, err := c.OpenDeviceWithVIDPID(0x8888, 0x0002)
-	if dev == nil {
-		t.Fatal("OpenDeviceWithVIDPID(0x8888, 0x0002): got nil device, need non-nil")
-	}
-	defer dev.Close()
+	dev, err := createFakeDevice(0x8888, 0x0002)
 	if err != nil {
-		t.Fatalf("OpenDeviceWithVIDPID(0x8888, 0x0002): %v", err)
+		t.Error(err)
 	}
 	dev.SetAutoDetach(true)
 	_, err = dev.Config(1)
 	if err == nil {
 		t.Fatalf("%s.Config(1) got nil, but want no nil because interface fails to detach", dev)
 	}
 }
+
+func TestActiveConfigNumFailure(t *testing.T) {
+	dev, err := createFakeDevice(0x8888, 0x0002)
+	if err != nil {
+		t.Error(err)
+	}
+
+	dev.handle = nil
+	_, err = dev.ActiveConfigNum()
+	if err == nil {
+		t.Fatalf("Error on failing ActiveConfigNum()")
+	}
+}
+
+func TestDeviceDescString(t *testing.T) {
+	dev, err := createFakeDevice(0x8888, 0x0002)
+	if err != nil {
+		t.Error(err)
+	}
+
+	expected := "1.2: 8888:0002 (available configs: [1])"
+	result := dev.Desc.String()
+
+	if expected != result {
+		t.Fatalf("Device String() failed, expeceted: \"%s\", got: \"%s\"", expected, result)
+	}
+}
+
+func createFakeDevice(vid ID, pid ID) (*Device, error) {
```

Comment:
I think you'll notice that as soon as you address the comments below, the helper is no longer useful, as it becomes:

func createFakeDevice(vid, pid ID) (*Device, error) {
  c := newContextWithImpl(newFakeLibusb())
  dev, err := c.OpenDeviceWithVIDPID(vid, pid)
  return dev, err
}

at which point it adds more lines of code than it saves. And at the same time it hides the opened context (c), which can no longer be closed by the caller. So I would recommend dropping this helper.